core/tests/tests_frontend.py

```diff
@@ -44,47 +45,63 @@ def test_register(self):
 
+    @suppress_warnings
+    def _test_global_preference_page_404(self, preference_name: str, page_url: str, user: TestSquireUser):
+        """ A global preference page should raise a http 404 if its corresponding global preference is not set """
         # Disable share URL
-        global_preferences['newsletter__share_link'] = ""
-        res = check_http_response(self, '/', 'get', TestAccountUser)
-        self.assertNotContains(res, reverse('core:newsletters'))
+        self.global_preferences[preference_name] = ""
+        check_http_response(self, page_url, 'get', user, response_status=404)
 
         # Enable share URL
-        global_preferences['newsletter__share_link'] = "https://www.example.com"
-        res = check_http_response(self, '/', 'get', TestAccountUser)
-        self.assertContains(res, reverse('core:newsletters'))
-
-    def test_homepage_message(self):
-        """ Tests if the homepage message is rendered properly """
-        global_preferences = global_preferences_registry.manager()
+        self.global_preferences[preference_name] = "https://www.example.com"
+        res = check_http_response(self, page_url, 'get', user)
+        self.assertContains(res, "https://www.example.com")
 
+    @suppress_warnings
+    def _test_global_preference_base_template_nav(self, preference_name: str, page_url: str, user: TestSquireUser):
+        """ A global preference tab should not be present in the base template nav if there is no share url """
         # Disable share URL
-        global_preferences['homepage__home_page_message'] = ""
-        res = check_http_response(self, '/', 'get', TestAccountUser)
-        self.assertNotContains(res, 'homepage_message')
+        self.global_preferences[preference_name] = ""
+        res = check_http_response(self, '/', 'get', user)
+        self.assertNotContains(res, page_url)
 
         # Enable share URL
-        global_preferences['homepage__home_page_message'] = "Test Homepage Message"
-        res = check_http_response(self, '/', 'get', TestAccountUser)
-        self.assertContains(res, "Test Homepage Message")
+        self.global_preferences[preference_name] = "https://www.example.com"
+        res = check_http_response(self, '/', 'get', user)
+        self.assertContains(res, page_url)
+
+    def test_newsletter(self):
+        """ Tests if the newsletter global preference is working """
+        self._test_global_preference_page_404('newsletter__share_link', reverse('core:newsletters'), TestMemberUser)
+        self._test_global_preference_base_template_nav('newsletter__share_link', reverse('core:newsletters'), TestMemberUser)
+
+    def test_downloads(self):
+        """ Tests if the downloads global preference is working """
+        self._test_global_preference_page_404('downloads__share_link', reverse('core:downloads'), TestMemberUser)
+        self._test_global_preference_base_template_nav('downloads__share_link', reverse('core:downloads'), TestMemberUser)
+
```

Comment:
I have to admit that I don’t know enough about the tests to say for sure if it covers everything, but it does pass and looks okay.

Reply:
CodeCov shows us which lines of changed code are not covered by tests directly within GitHub. For example, see CodeCov's comment in #247:

<details>

![image](https://user-images.githubusercontent.com/34304046/189606580-461355e9-40c0-4500-be30-e652cda85280.png)
</details>

Of course, coverage doesn't tell us anything about the quality (or even correctness) of the tests, which is what these reviews are for. It seems like the diff doesn't really show well what actually changed here though, so let me walk you through it.

1. Although it shows up as changed, `def test_homepage_message(self):` wasn't actually touched at all.
2. For both the newsletter and downloads page, we want to test two things:
   1. Their individual pages `/newsletters` and `/downloads` should raise a HTTP 404 if no share link is set in the global preferences, and they should be accessible if such a share link is set.
   2. Links to these individual pages should not be present in the navigation bar if such a link is not set, and they should be linked to if it is set.
3. These two tests were already present for newsletters previously. I could have copy-pasted and adapted those tests for the Downloads and called it a day, but that would result in code duplication. Instead, I extracted two general methods based on the two newsletter tests, which get called with different parameters for Newsletters and Downloads.
`def _test_global_preference_page_404(..)` tests condition "i" mentioned above, whereas `def _test_global_preference_base_template_nav(..)` tests condition "ii". Tests were also moved to a different Test-class `NavGlobalPreferenceTest` for better organisation.

To verify that this change was made correctly, you could:
1. Check whether the old tests for newsletters were correctly generalised into the checks `def _test_global_preference_page_404(..)` and `def _test_global_preference_base_template(..)`
2. Check that the right parameters are used when calling these general methods.

Reply:
> Of course, coverage doesn't tell us anything about the quality (or even correctness) of the tests, which is what these reviews are for.

Yeah, I meant it that way.  
And I wasn’t familiar enough to assess that properly.  
I mean the tests looked they made sense, and I got most of what you said in your comment already, but I missed `page_url` thing @DutcherNL pointed out, which I agree on.

Reply:
I agree the comments could have been a lot more helpful there. I've improved them in the latest commit 👍